libsmi/tools/smilint.1.in
.\"
.\" $Id: smilint.1.in 1174 2004-08-10 10:58:12Z strauss $
.\"
.TH smilint 1  "August 10, 2004" "IBR" "SMI Tools"
.SH NAME
smilint \- syntax and semantic checks of SMIv1/v2 and SPPI modules
.SH SYNOPSIS
.B smilint
[
.B "-Vhersm"
] [
.BI "-c " file
] [
.BI "-p " module
] [
.BI "-l " level
] [
.BI "-i " error-pattern
]
.I "module(s)"
.SH DESCRIPTION
The \fBsmilint\fP program is used to check MIB or PIB modules for syntax
errors and semantics at some degree.  SMIv1/v2 style
MIB modules as well as SPPI PIB modules are supported.
.PP
The rules that \fBsmilint\fP is based on are taken from RFC 1155, RFC
1212 and RFC 1215 for SMIv1, RFCs 2578-2580 for SMIv2, RFC 3159 for SPPI.
.\" and the specification of SMIng that has been a research project
.\" within the Network Management Research Group of the IRTF.
.SH OPTIONS
.TP
\fB-V, --version\fP
Show the smilint version and exit.
.TP
\fB-h, --help\fP
Show a help text and exit.
.TP
\fB-e, --error-list\fP
Show a list of all known error messages and exit. Error messages can
have associated tags, shown in braces at the end of each line. The
tags can be used with the -i option to ignore certain error messages.
.TP
\fB-r, --recursive\fP
Report errors and warnings also for recursively imported modules. 
.TP
\fB-s, --severity\fP
Show the error severity in brackets before error messages.
.TP
\fB-m, --error-names\fP
Show the error names in braces before error messages.
.TP
\fB-c \fIfile\fB, --config=\fIfile\fP
Read \fIfile\fP instead of any other (global and user)
configuration file.
.TP
\fB-p \fImodule\fB, --preload=\fImodule\fP
Preload the module \fImodule\fP before reading the main
module(s). This may be helpful if an incomplete main module misses to
import some definitions.
.TP
\fB-l \fIlevel\fB, --level=\fIlevel\fP
Report errors and warnings up to the given severity \fIlevel\fP.
See below for a description of the error levels. The default error
level is 3.
.TP
\fB-i \fIprefix\fB, --ignore=\fIprefix\fP
Ignore all errors that have a tag which matches \fIprefix\fP.
A list of error tags can be retrieved by calling smilint with the
-e option.
.TP
.I module(s)
These are the modules to be checked. If a module argument represents a
path name (identified by containing at least one dot or slash character),
this is assumed to be the exact file to read. Otherwise, if a module is
identified by its plain module name, it is searched according to libsmi
internal rules. See \fBsmi_config(3)\fP for more details.
.SH "ERROR AND WARNING LEVELS"
All generated error and warning messages have an associated severity level.
The actual severity levels are:
.IP 0 3
Internal error, no recovery possible. Examples are memory allocation
failures. Errors of this level usually cause the application to abort.
.IP 1 3
Major SMI/SPPI error, recovery somehow possible but may lead to severe
problems. Examples are lexically unexpected characters or unknown
keywords. Errors of this kind usually lead to follow-on errors.
.IP 2 3
SMI/SPPI error which is probably tolerated by some
implementations. Examples are MIB/PIB modules which mix constructs
from different SMI/SPPI versions.
.IP 3 3 
SMI/SPPI error which is likely tolerated by many
implementations. Examples are misplaced SMIv2 MODULE-IDENTITY
invocations or SMIv2 textual conventions derived from other textual
conventions.
.IP 4 3
Something which is not strictly an error but which is recommended to
be changed. Warnings of this level are usually considered during MIB
reviews.
.IP 5 3
Something that is basically correct but might be problematic in
certain environments or usage scenarios. Examples are warnings that
identifiers only differ in case or that type definitions are not used
within the defining module.
.IP 6 3
Messages of this level are auxiliary notices. Examples are messages
that point to a previous definition in case of a redefinition.
.PP
Higher levels are currently not used and lead to the same effects as
level 6 does. Note that errors up to level 3 are errors violating the
specifications and must be fixed by the responsible author. The
warnings generated with level 4 should be considered during normal
MIB/PIB reviews.
.SH "EXAMPLE"
This example checks the file RMON2-MIB in the current directory (note
that the `./' prefix ensures this). The error level is raised to 6
and warnings that claim about identifier names that exceed a length
of 32 characters are suppressed.
.nf

  $ smilint -l 6 -i namelength-32 ./RMON2-MIB
  ./RMON2-MIB:3935: unexpected type restriction
  ./RMON2-MIB:3936: unexpected type restriction
  ./RMON2-MIB:3937: unexpected type restriction
  ./RMON2-MIB:3938: unexpected type restriction
  ./RMON2-MIB:3939: unexpected type restriction
  ./RMON2-MIB:3940: unexpected type restriction
  ./RMON2-MIB:4164: scalar object must not have a `read-create' access value

.fi
.SH "SEE ALSO"
The 
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/projects/libsmi/" "."
Other commonly used MIB checkers are
.BR mosy "(1) and "
.BR smicng "(1)."
.SH "AUTHORS"
(C) 1999-2004 F. Strauss, TU Braunschweig, Germany <strauss@ibr.cs.tu-bs.de>
.br
(C) 1999-2002 J. Schoenwaelder, TU Braunschweig, Germany <schoenw@ibr.cs.tu-bs.de>
.br
(C) 2002-2003 J. Schoenwaelder, University of Osnabrueck, Germany
.br
(C) 2003-2004 J. Schoenwaelder, International University Bremen, Germany
.br
(C) 2001-2002 T. Klie, TU Braunschweig, Germany <tklie@ibr.cs.tu-bs.de>
.br
(C) 2002 M. Bunkus, TU Braunschweig, Germany <bunkus@ibr.cs.tu-bs.de>
.br
and contributions by many other people.
.br


qwt/CHANGES
Release 5.1.0
===================

Changes
-------
1) QwtSymbol::copy introduced
   Now it is possible to use derived symbol classes for curves
2) QwtPlotScaleItem introduced
   A new type of plot item for displaying axes on the canvas
3) QwtClipper added
   A collection of clipping algos
4) Using DBL_EPSILON
   This change allows smaller intervals for sliders/dials
5) QwtPanner
   setOrientation() added.
6) QwtPlot
   axisStepSize() added
   clear is virtual now
7) QwtPlotPrintFilter
   PrintCanvasBackground splitted into PrintBackground, PrintFrameWithScales
8) QwtPlotZoomer
   setZoomStack() added
9) Changes for the QwtPolar package
   QwtLegendItemManager introduced
   QwtMagnifier introduced
10)Suffix rules added in qwtconfig.pri for different targets for 
   debug/release builds.

Bug Fixes
---------
1. QwtAbstractScaleDraw::setAbstractScaleDraw
   Reinitialization problem fixed
2. QwtLegendItem
   key event handlers fixed
3. QwtPicker
   solaris-cc compiler problem fixed
4. Inaccurate mapping of scale to widget coordinates fixed
5. QwtPlotCurve::draw
   Updates for Qt 4.3 added
6. QwtPlotLayout
   AlignToCanvas layout calculation fixed
7. QwtPlot::print
   Workaround for a QPen initialization problem, 
   when printing to Pdf, added
8. QwtText
   Layout of rich text documents fixed
9. Designer
   Handling of QwtScaleWidget fixed
10. realtime example
   Qt::WA_PaintOutsidePaintEvent added, ScrollZoomer fixed
11. Several others I have forgotten

Release 5.0.2
===================

Bug Fixes
---------
1. QwtPlotCurve::Xfy curve type fixed
2. Memory leak in QwtLegend fixed
3. Vertical alignment of rich texts fixed
4. Workaround for a Qt4 bug added, that produces horrible performance
   when painting curves with a pen width > 1.
5. Background for the tracker text of QwtPickers fixed.
   Improved (faster + better rendered texts) implementation of
   painting tracker texts, using capabilities of Qt >= 4.3.
6. QwtArrowButton/QwtCounter: workaround for layout bug ( Qt < 4.3 )
   of the Cleanlook style added.
7. A couple of minor fixes

Changes
-------
1. QSvgGenerator added to the bode example

Release 5.0.1
===================

Changes
-------
1.  A couple of problems, when building Qwt fixed.
2.  Displaying Rich Text with Qt 4.x fixed

Release 5.0.0
===================

Platforms
---------
Support of Qt3 and Qt4. Qt2 is not supported any longer.

Key features
------------
1.  Redesign of plot items. Makes it much easier to develop
    individual items.
2.  Redesign of the scale classes. All calculations are
    collected in scale engines, where the application can
    implement it´s own (f.e log2, or date scales). Now it´s
    also possible to have individual and completely irregular scales
3.  Redesign of the QwtText classes. The MathML renderer of
    the Qt4 solutions package is embedded. 
    work for all expressions/situations.
4.  New classes for navigating: QwtPanner, QwtMaginfier
5.  Spectrogram/Contour plots  and other classes for displaying
    raster data added.

Changes
-------
5.0.0 is by far the release with the most changes in the history of Qwt
- too many to make list.


Release 4.2.0/0.4.2
===================

License
--------
A couple of exceptions to the LGPL with the intention to allow static
linking with commercial applications. See COPYING.

Key features: 
-------------
1.  Designer plugin 
2.  Rich Text support ( f.e. E = m * c<sup>2</sup> ) added.
3.  QwtDial class family added (QwtDial, QwtCompass, QwtAnalogClock, ...)
4.  QwtPicker class family added. Includes QwtPlotZoomer, a complete
    implementation of recursive zooming.
5.  Device metrics independent printing of QwtPlot. (QPrinter::HighResolution)
6.  QwtPlot::setCurveBrush(), QwtCurve::setBrush() added. The area
    between curve and baseline will be filled with this brush.
7.  Rotation of axis tick labels added. Very useful for axis with long
    labels like time scales ...
8.  Added a new abstract QwtData class to plot data from almost any type of
    container class.  
9.  QwtDoublePoint, QwtDoubleSize, QwtDoubleRect double counterparts
    for QPoint, QSize, QRect.
10. First steps to support Qtopia. All examples can be compiled and started
    in the qvfb emulator.


Changes:
---------
1.  Rewrite of QwtLegend/QwtLegendItem (no QTable anymore)
2.  Each plot item will be painted, even if one of the axis it is 
    attached to is disabled. (like in all other releases beside 0.4.1)
3.  Code for double buffering moved to a new class QwtPaintBuffer.
    Double buffering can be enabled/disabled now. 
4.  QwtPainter, QwtMetricsMap, QwtLayoutMetrics added
    Hide paint device metrics dependencies.
5.  Layout code rewritten and moved to a new class QwtPlotLayout
    New layout options canvasMargin(), alignCanvasToScales()
6.  QwtPlot: sizeHint() != minimumSizeHint()
9.  Internal plot data are private again. A couple of get methods 
    added instead.
10. canvas repaints triggered by paint events. Enables event filtering
11. QwtPlot::drawCanvasItems added. In opposite to QwtPlot::drawCanvas
    it is used by the printing code too.
12. qwtMax, qwtMin, qwtInt mapped to QMAX, QMIN, qRound from qglobal.h
13. operator= for plot item classes changed.
14. readOnly property added for sliders.
15. valid flag added for QwtDblRange
16. QwtCounter wrap around policy: a counter under- or overflow sets
    focus to the smallest up/down button and disables counting.
    A space bar keypress release event re-enables counting.
17. QwtPushButton added. A class that adds rich text and alignments
    features to QPushButton, like they are used in QLabel
18. Clipped painting code moved from QwtCurve to QwtPainter/QwtRect
19. Canvas cache added to optimize trivial repaints.
20. QwtPlot::drawCurve added for incremental curve data
21. QwtSliderBase, readOnly, isValid added
22. Added filtering of the colors of the title and scales to QwtPrintFilter.
23. Support of QT_NO_CAST_ASII and QT_NO_COMPAT added

libsmi/tools/smilint.c
/*
 * smilint.c --
 *
 *      MIB module checker main program.
 *
 * Copyright (c) 1999 Frank Strauss, Technical University of Braunschweig.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smilint.c 1673 2010-06-24 10:09:36Z schoenw $
 */

#include <config.h>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_WIN_H
#include "win.h"
#endif

#include "smi.h"
#include "shhopt.h"



/*
 * These are functions that are not officially exported by the libsmi.
 * See the original prototype definitions in lib/error.h.
 */

extern int smiGetErrorSeverity(int id);
extern char* smiGetErrorTag(int id);
extern char* smiGetErrorMsg(int id);
extern char* smiGetErrorDescription(int id);


static int mFlag = 0;	/* show the name for error messages */
static int sFlag = 0;	/* show the severity for error messages */
static int eFlag = 0;	/* print the list of possible error messages */
static int flags;


typedef struct Error {
    int id;
    int severity;
    char *tag;
    char *msg;
    char *description;
    int used;
} Error;


static Error *errors = NULL;


static void fold(FILE *f, int indent, const char *msg)
{
    const char *p, *s;

    if (! msg) {
	fprintf(f, "\n");
	return;
    }

    for (s = msg; *s; s++) {
	for (p = s; *p && *p != '\n'; p++) ;
	if (*p) {
	    fprintf(f, "%.*s\n%*s", (int) (p - s), s, indent, "");
	    s = p;
	} else {
	    fprintf(f, "%.*s\n", (int) (p - s), s);
	    break;
	}
    }
}



static int compare(const void *v1, const void *v2)
{
    Error *err1 = (Error *) v1;
    Error *err2 = (Error *) v2;

    if (err1->severity < err2->severity) {
	return -1;
    }
    if (err1->severity > err2->severity) {
	return 1;
    }
    return strcmp(err1->msg, err2->msg);
}



static Error* errors_new()
{
    int i, cnt;
    Error *errors;
    
    for (cnt = 0; smiGetErrorSeverity(cnt) >= 0; cnt++) ;
    
    errors = malloc((cnt + 1) * sizeof(Error));
    if (! errors) {
	fprintf(stderr, "smilint: malloc failed - running out of memory\n");
	exit(1);
    }
    memset(errors, 0, (cnt + 1) * sizeof(Error));

    for (i = 0; i < cnt; i++) {
	errors[i].id = i;
	errors[i].severity = smiGetErrorSeverity(i);
	errors[i].tag = smiGetErrorTag(i);
	errors[i].msg = smiGetErrorMsg(i);
	errors[i].description = smiGetErrorDescription(i);
    }

    qsort(errors, cnt, sizeof(Error), compare);

    return errors;
}


static void display_one(FILE *f, Error *error)
{
    const int indent = 12;
    char *type, *tag;

    type = (error->severity <= 3) ? "Error:" : "Warning:";
    tag = (error->tag && strlen(error->tag))
	? error->tag : "<xxx-missing-xxx>";
    fprintf(f, "%-*s %s (level %d%s)\n",
	    indent, type, tag, error->severity & 127,
	    error->severity & 128 ? ", ignored" : "");
    fprintf(f, "%-*s %s\n", indent, "Message:",
	    error->msg ? error->msg : "");
    if (error->description) {
	fprintf(f, "%-*s ", indent, "Description:");
	fold(f, indent + 1, error->description);
    }
}


static void display_all(Error *errors)
{
    int i;
    
    for (i = 0; errors[i].msg; i++) {
	if (i) printf("\n");
	display_one(stdout, errors + i);
    }
}



static void display_used(Error *errors)
{
    int i, n;

    for (i = 0, n = 0; errors[i].msg; i++) {
	if (errors[i].used && errors[i].description) n++;
    }

    if (! n) {
	return;
    }

    fprintf(stderr,
	    "\nAdditional descriptions of some error/warning messages:\n"); 

    for (i = 0; errors[i].msg; i++) {
	if (! errors[i].used || !errors[i].description) continue;
	if (i) fprintf(stderr, "\n");
	display_one(stderr, errors + i);
    }
}



static void usage()
{
    fprintf(stderr,
	    "Usage: smilint [options] [module or path ...]\n"
	    "  -V, --version         show version and license information\n"
	    "  -h, --help            show usage information\n"
	    "  -c, --config=file     load a specific configuration file\n"
	    "  -p, --preload=module  preload <module>\n"
	    "  -e, --error-list      print list of known error messages\n"
	    "  -m, --error-names     print the name of errors in braces\n"
	    "  -s, --severity        print the severity of errors in brackets\n"
	    "  -r, --recursive       print errors also for imported modules\n"
	    "  -l, --level=level     set maximum level of errors and warnings\n"
	    "  -i, --ignore=prefix   ignore errors matching prefix pattern\n"
	    "  -I, --noignore=prefix do not ignore errors matching prefix pattern\n");
}



static void help() { usage(); exit(0); }
static void version() { printf("smilint " SMI_VERSION_STRING "\n"); exit(0); }
static void config(char *filename) { smiReadConfig(filename, "smilint"); }
static void preload(char *module) { smiLoadModule(module); }
static void recursive() { flags |= SMI_FLAG_RECURSIVE; smiSetFlags(flags); }
static void level(int lev) { smiSetErrorLevel(lev); }
static void ignore(char *ign) { smiSetSeverity(ign, 128); }
static void noignore(char *ign) { smiSetSeverity(ign, -1); }



static void
errorHandler(char *path, int line, int severity, char *msg, char *tag)
{
    int i;
    
    if (path) {
	fprintf(stderr, "%s:%d: ", path, line);
    }
    if (sFlag) {
	fprintf(stderr, "[%d] ", severity);
    }
    if (mFlag) {
	fprintf(stderr, "{%s} ", tag);
    }
    switch (severity) {
    case 4:
    case 5:
	fprintf(stderr, "warning: ");
	break;
    case 6:	
	fprintf(stderr, "info: ");
	break;
    }
    fprintf(stderr, "%s\n", msg);

    if (severity <= 0) {
	exit(1);
    }

    /* If we are supposed to generate error descriptions, locate this
     * error in our error list and increment its usage counter. Note
     * that we assume that error tags are unique (and we should better
     * check for this somewhere). */

    if (errors) {
	for (i = 0; errors[i].msg; i++) {
	    if (strcmp(errors[i].tag, tag) == 0) {
		errors[i].used++;
		break;
	    }
	}
    }
}



int main(int argc, char *argv[])
{
    int i;

    static optStruct opt[] = {
	/* short long              type        var/func       special       */
	{ 'h', "help",           OPT_FLAG,   help,          OPT_CALLFUNC },
	{ 'V', "version",        OPT_FLAG,   version,       OPT_CALLFUNC },
	{ 'c', "config",         OPT_STRING, config,        OPT_CALLFUNC },
	{ 'p', "preload",        OPT_STRING, preload,       OPT_CALLFUNC },
	{ 'e', "error-list",     OPT_FLAG,   &eFlag,        0 },
	{ 'm', "error-names",    OPT_FLAG,   &mFlag,        0 },
	{ 's', "severity",       OPT_FLAG,   &sFlag,        0 },
	{ 'r', "recursive",      OPT_FLAG,   recursive,     OPT_CALLFUNC },
	{ 'l', "level",          OPT_INT,    level,         OPT_CALLFUNC },
	{ 'i', "ignore",         OPT_STRING, ignore,        OPT_CALLFUNC },
	{ 'I', "noignore",       OPT_STRING, noignore,      OPT_CALLFUNC },
	{ 0, 0, OPT_END, 0, 0 }  /* no more options */
    };
    
    for (i = 1; i < argc; i++)
	if ((strstr(argv[i], "-c") == argv[i]) ||
	    (strstr(argv[i], "--config") == argv[i])) break;
    if (i == argc) 
	smiInit("smilint");
    else
	smiInit(NULL);

    flags = smiGetFlags();
    flags |= SMI_FLAG_ERRORS;
    flags |= SMI_FLAG_NODESCR;
    smiSetFlags(flags);

    optParseOptions(&argc, argv, opt, 0);

    if (eFlag) {
	mFlag = 1;
	errors = errors_new();
    }

    if (sFlag || mFlag) {
	smiSetErrorHandler(errorHandler);
    }

    if (eFlag && argc == 1) {
	if (errors) {
	    display_all(errors);
	    free(errors);
	}
	smiExit();
	return 0;
    }
    
    for (i = 1; i < argc; i++) {
	if (smiLoadModule(argv[i]) == NULL) {
	    fprintf(stderr, "smilint: cannot locate module `%s'\n",
		    argv[i]);
	    smiExit();
	    exit(1);
	}
    }

    if (eFlag) {
	if (errors) {
	    display_used(errors);
	    free(errors);
	}
    }

    smiExit();

    return 0;
}
